### PR TITLE
[operator] Reconcile `seedmanagement.gardener.cloud/v1alpha1.Gardenlet` object when there is no corresponding `Seed` yet

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement

**What this PR does / why we need it**:
When there is no corresponding `Seed` resource yet, `gardener-operator` reconciles the `seedmanagement.gardener.cloud/v1alpha1.Gardenlet` resource. This facilitates the initial deployment of `gardenlet`s into unmanaged seed clusters. It helps human operators to set up such seeds declaratively (previously, this was a manual process requiring them to create a bootstrap token, craft a bootstrap kubeconfig, prepare Helm chart values, and run the chart deployment).

After `gardenlet` created the `Seed`, `gardener-operator` will never act on it again, and `gardenlet` relies on the self-upgrade procedure introduced as part of  #9830.

The local operator setup w/ seeds was adapted to not deploy the `gardenlet` Helm chart anymore, but create the corresponding `seedmanagement.gardener.cloud/v1alpha1.Gardenlet` resource instead.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/8802

**Special notes for your reviewer**:
/cc @ScheererJ 

**Release note**:

```feature operator
`gardener-operator` is now capable of deploying `gardenlet`s into unmanaged seed clusters. Read more about it [here](https://github.com/gardener/gardener/tree/master/docs/deployment/deploy_gardenlet_via_operator.md).
```

### CODE REVIEW

File: docs/deployment/deploy_gardenlet.md [hunk 1]
```diff
@@ -4,12 +4,15 @@ Gardenlets act as decentralized agents to manage the shoot clusters of a seed cl
...
+
+The following documents describe the scenarios:
+
+1. If you have not registered a seed cluster yet (thus, you need to deploy a first, so-called "unmanaged seed"), your approach depends on how you deployed the Gardener control plane:
+   1. **Via [`gardener-operator`](../concepts/operator.md)**:
+      1. If you want to register the same cluster in which `gardener-operator` runs, or if you want to register another cluster that is reachable (network-wise) for `gardener-operator`, you can follow [Deploy gardenlet via `gardener-operator`](deploy_gardenlet_via_operator.md).
+      2. If you want to register a cluster that is not reachable (network-wise) (e.g., because it runs behind a firewall), you can follow [Deploy a gardenlet Manually](deploy_gardenlet_manually.md). 
+   2. **Via [`gardener/controlplane` Helm chart](../../charts/gardener/controlplane)**: You can follow [Deploy a gardenlet Manually](deploy_gardenlet_manually.md).
+2. If you already have a seed cluster, and you want to deploy further seed clusters (so-called "managed seeds"), you can follow [Deploy a gardenlet Automatically](deploy_gardenlet_automatically.md).
```
Comment (ScheererJ) [inline docs/deployment/deploy_gardenlet.md]: The nesting feels a bit deep with these three layers. Maybe less would look/read better.

For example:

```
1. Register first seed cluster when Gardener control plane was deployed via operator and directly reachable...
2. Register first seed cluster when Gardener control plane was deployed via Helm chart or is not directly reachable...
3. Register additional managed seed clusters...
```

However, this is mostly personal preference.

File: docs/deployment/deploy_gardenlet_via_operator.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+Using this method only works when [`gardener-operator`](../concepts/operator.md) is managing the garden cluster.
+If you have used the [`gardener/controlplane` Helm chart](../../charts/gardener/controlplane) for the deployment of the Gardener control plane, please refer to [this document](deploy_gardenlet_manually.md).
+
+> [!TIP]
+> The initial seed cluster can be the garden cluster itself, but for better separation of concerns, it is recommended to only register other clusters as seeds.
+
+## Deployment of gardenlets
+
+Using this method, `gardener-operator` is only taking care of the very first deployment of gardenlet.
+Once running, the gardenlet leverage the [self upgrade](deploy_gardenlet_manually.md#self-upgrades) strategy in order to keep themselves up-to-date.
```
Comment (ScheererJ) [inline docs/deployment/deploy_gardenlet_via_operator.md]:
```suggestion
Once running, the gardenlet leverages the [self upgrade](deploy_gardenlet_manually.md#self-upgrades) strategy in order to keep itself up-to-date.
```

Alternatively use `gardenlet`s instead.

File: docs/deployment/deploy_gardenlet_via_operator.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+              namespace: garden
+            type: local
+        ingress:
+          controller:
+            kind: nginx
+          domain: ingress.local.seed.local.gardener.cloud
+        networks:
+          nodes: 172.18.0.0/16
+          pods: 10.1.0.0/16
+          services: 10.2.0.0/16
```
Comment (ScheererJ) [inline docs/deployment/deploy_gardenlet_via_operator.md:84]: Is this required if deploying into the garden runtime cluster? Does `gardener-operator` not have that knowledge already? In other words, can we reduce the amount of configuration necessary for the "simple" case of garden runtime cluster = first seed cluster?

File: docs/deployment/deploy_gardenlet_via_operator.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+          excessCapacityReservation:
+            enabled: false
+          scheduling:
+            visible: true
+          verticalPodAutoscaler:
+            enabled: true
+```
+
+This causes `gardener-operator` to deploy gardenlet to the same cluster where it is running.
+Once it came up, gardenlet will create a `Seed` resource with the same name and uses the `Gardenlet` resource for self-upgrades (see [this document](deploy_gardenlet_manually.md#self-upgrades)).
```
Comment (ScheererJ) [inline docs/deployment/deploy_gardenlet_via_operator.md]:
```suggestion
Once it comes up, gardenlet will create a `Seed` resource with the same name and uses the `Gardenlet` resource for self-upgrades (see [this document](deploy_gardenlet_manually.md#self-upgrades)).
```

File: docs/deployment/deploy_gardenlet_via_operator.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+### Remote Clusters
+
+If you want it to deploy gardenlet into some other cluster, create a kubeconfig `Secret` and reference it in the `Gardenlet` resource:
```
Comment (ScheererJ) [inline docs/deployment/deploy_gardenlet_via_operator.md]:
```suggestion
If you want `gardener-operator` to deploy gardenlet into some other cluster, create a kubeconfig `Secret` and reference it in the `Gardenlet` resource:
```

File: example/operator/30-gardenlet.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+#   image: # usually not needed since image is implicitly defined by above helm chart
+#     repository: gardenlet-image:tag
+#     tag: latest
+#     pullPolicy: IfNotPresent
+#   resources:
+#     requests:
+#       cpu: 100m
+#       memory: 100Mi
+#     limits:
+#       cpu: 2000m
```
Comment (ScheererJ) [inline example/operator/30-gardenlet.yaml]: CPU limits are a very mixed bag. Should we advertise this here?

File: pkg/controller/gardenletdeployer/actuator.go [hunk 7]
```diff
@@ -49,53 +49,29 @@ const (
...
-		gardenConfig:         gardenConfig,
-		gardenAPIReader:      gardenAPIReader,
-		gardenClient:         gardenClient,
-		seedClient:           seedClient,
-		shootClientMap:       shootClientMap,
-		clock:                clock,
-		vp:                   vp,
-		recorder:             recorder,
-		gardenNamespaceShoot: gardenNamespaceShoot,
-	}
```
Comment (ScheererJ) [inline pkg/controller/gardenletdeployer/actuator.go]: Any particular reason why the constructor was removed?

File: pkg/operator/controller/gardenlet/add.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+// OperatorResponsiblePredicate is a predicate for checking whether the Seed object has already been created for the
+// Gardenlet resource, and whether the kubeconfig secret ref has been removed.
+func (r *Reconciler) OperatorResponsiblePredicate(ctx context.Context) predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		gardenlet, ok := obj.(*seedmanagementv1alpha1.Gardenlet)
+		if !ok {
+			return false
+		}
+		return r.seedDoesNotExist(ctx, gardenlet) || gardenlet.Spec.KubeconfigSecretRef != nil
```
Comment (ScheererJ) [inline pkg/operator/controller/gardenlet/add.go:93]: I am surprised to see the check on the `KubeconfigSecretRef` here. It looks like we will not get into `r.reconcile(...)` due to the short-cut in `r.Reconcile(...)` (`if !r.seedDoesNotExist(ctx, gardenlet) {`).

How is this supposed to work?

File: hack/test-e2e-local.sh [hunk 9]
```diff
@@ -55,6 +55,12 @@ if [[ "$TYPE" == "operator" ]] || [[ "$TYPE" == "operator-seed" ]]; then
       exit 1
     fi
   fi
+
+  if [[ "$TYPE" == "operator-seed" ]]; then
+    # /etc/hosts must be updated before garden can be created
+    echo "> Deploying Garden and Soil"
+    make operator-seed-up
+  fi
```
Comment (ScheererJ) [inline hack/test-e2e-local.sh]: Is the comment still true? If so this has moved to a wrong position as it is now before the `/etc/hosts` manipulation.

**Review by ScheererJ — COMMENTED** [review]
Thanks a lot for this great improvement in terms of setting up an initial Gardener landscape.

File: pkg/operator/controller/gardenlet/add.go (hunk 8 above)
Comment (rfranzke) [reply pkg/operator/controller/gardenlet/add.go:93]: Nice catch :) I added this check in the `Reconcile` function only later and forgot about this predicate. Adapted now.

File: hack/test-e2e-local.sh (hunk 9 above)
Comment (rfranzke) [reply hack/test-e2e-local.sh]: It's still true - a few lines above, the `/etc/hosts` file is manipulated with the garden-specific domain names.

File: pkg/controller/gardenletdeployer/actuator.go (hunk 7 above)
Comment (rfranzke) [reply pkg/controller/gardenletdeployer/actuator.go]: I didn't find it providing particular value, so I removed it. The `Actuator` struct is now exposed and its attributes are set directly.

File: docs/deployment/deploy_gardenlet_via_operator.md (hunk 3 above)
Comment (rfranzke) [reply docs/deployment/deploy_gardenlet_via_operator.md:84]: Maybe, but this is something that we should probably do separately. `gardener-operator` would need a webhook on `Seed`s and default things that are already known by the `Garden` resource. However, it's a very neat/special case, so not sure if it's generally worth the effort/code/complexity.

File: example/55-gardenlet.yaml [hunk 10]
```diff
@@ -37,3 +37,7 @@ spec:
     apiVersion: gardenlet.config.gardener.cloud/v1alpha1
     kind: GardenletConfiguration
     <See `20-componentconfig-gardenlet.yaml` for more details>
+# For the very first deployment of a gardenlet via gardener-operator, you can specify a secret containing a kubeconfig
+# for the target cluster. gardener-operator will then attempt to deploy gardenlet into it.
+# kubeconfigSecretRef:
+#   name: <name-to-secret-containing-kubeconfig>
```
Comment (acumino) [inline example/55-gardenlet.yaml:43]:
```suggestion
  # For the very first deployment of a gardenlet via gardener-operator, you can specify a secret containing a kubeconfig
  # for the target cluster. gardener-operator will then attempt to deploy gardenlet into it.
  # kubeconfigSecretRef:
  #   name: <name-to-secret-containing-kubeconfig>
```

File: pkg/apis/seedmanagement/v1alpha1/types_gardenlet.go [hunk 11]
```diff
@@ -49,6 +50,12 @@ type GardenletSpec struct {
 	// Config is the GardenletConfiguration used to configure gardenlet.
 	// +optional
 	Config runtime.RawExtension `json:"config,omitempty" protobuf:"bytes,2,opt,name=config"`
+	// KubeconfigSecretRef is a reference to a secret containing a kubeconfig for the cluster to which gardenlet should
+	// be deployed. This is only used by gardener-operator for a very first gardenlet deployment. After that, gardenlet
+	// will continuously upgrade itself. If this field is empty, gardener-operator deploys it into its own runtime
+	// cluster.
+	// +optional
+	KubeconfigSecretRef *corev1.LocalObjectReference `json:"kubeconfigSecretRef,omitempty" protobuf:"bytes,3,opt,name=kubeconfigSecretRef"`
```
Comment (acumino) [inline pkg/apis/seedmanagement/v1alpha1/types_gardenlet.go:58]: This is not added in the internal version, is there some specific reason for it?

File: charts/gardener/operator/templates/_helpers.tpl [hunk 12]
```diff
@@ -95,6 +95,12 @@ config.yaml: |
 {{ toYaml .Values.config.controllers.gardenCare.conditionThresholds | indent 6 }}
       {{- end }}
     {{- end }}
+    {{- if .Values.config.controllers.gardenletDeployer }}
+    gardenletDeployer:
+      {{- if .Values.config.controllers.gardenletDeployer.concurrentSyncs }}
+      concurrentSyncs: {{ .Values.config.controllers.networkPolicy.concurrentSyncs }}
```
Comment (acumino) [inline charts/gardener/operator/templates/_helpers.tpl]:
```suggestion
      concurrentSyncs: {{ .Values.config.controllers.gardenletDeployer.concurrentSyncs }}
```

**Review by acumino — CHANGES_REQUESTED** [review]

File: example/55-gardenlet.yaml (hunk 10 above)
Comment (rfranzke) [reply example/55-gardenlet.yaml:43]: This indentation is intentional, what's the concern/suggestion?

Comment (acumino) [reply example/55-gardenlet.yaml:43]: My idea was to first add the fields and then comment it out, so when we uncomment it, it is properly indented

Comment (rfranzke) [reply example/55-gardenlet.yaml:43]: I understand, but it is not done like this for most other examples in this repo, hence, I aligned with this. The current approach has the benefit that you can directly see where this field belongs. In the end, it's a question of consistency.

File: hack/test-e2e-local.sh [hunk 13]
```diff
@@ -57,7 +57,8 @@ if [[ "$TYPE" == "operator" ]] || [[ "$TYPE" == "operator-seed" ]]; then
   fi
 
   if [[ "$TYPE" == "operator-seed" ]]; then
-    # /etc/hosts must be updated before garden can be created
+    # /etc/hosts must have been updated before garden can be created (otherwise, we could put this command to the
+    # hack/ci-e2e-kind-operator-seed.sh script.
     echo "> Deploying Garden and Soil"
     make operator-seed-up
   fi
```
Comment (ScheererJ) [inline hack/test-e2e-local.sh]: `make operator-seed-up` creates the `Garden` if I understand it correctly. However, `/etc/hosts` is updated below. Therefore, I am still confused by the comment.
(In addition to that, there is a closing parenthesis missing.)

**Pull request merged**